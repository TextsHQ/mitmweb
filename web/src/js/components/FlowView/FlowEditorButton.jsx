import React, { PropTypes, Component } from 'react'
import { connect } from 'react-redux'

import {closeFlowEditor} from '../../ducks/ui.js'
import {openFlowEditor} from '../../ducks/ui.js'

// FlowEditorButton.propTypes = {
//     isFlowEditorOpen: PropTypes.bool.isRequired,
//     content: PropTypes.string.isRequired,
//     onContentChange: PropTypes.func.isRequired
// }

class FlowEditorButton extends Component{
     static propTypes = {
        isFlowEditorOpen: PropTypes.bool.isRequired,
        content: PropTypes.string.isRequired,
        onContentChange: PropTypes.func.isRequired
     }

    render(){
        let { isFlowEditorOpen, closeFlowEditor, openFlowEditor,  onContentChange, content } = this.props
        return (
            <div className="edit-flow-container">
                {isFlowEditorOpen ?
                    <a className="edit-flow" onClick={() => {onContentChange(content); closeFlowEditor()}}>
                        <i className="fa fa-check"/>
                    </a>
                    :
                    <a  className="edit-flow" onClick={() => openFlowEditor()}>
                        <i className="fa fa-pencil"/>
                    </a>
                }
            </div>
        )
    }
}

export default connect(
    state => ({
        isFlowEditorOpen: state.ui.isFlowEditorOpen,
        content: state.flows.modifiedFlow.content
    }),
    {
        closeFlowEditor,
        openFlowEditor

    }
)(FlowEditorButton)
