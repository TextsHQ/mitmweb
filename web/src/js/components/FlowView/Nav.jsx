import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import classnames from 'classnames'
import * as flowsActions from '../../ducks/flows'

NavAction.propTypes = {
    icon: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
}

function NavAction({ icon, title, onClick }) {
    return (
        <a title={title}
            href="#"
            className="nav-action"
            onClick={event => {
                event.preventDefault()
                onClick(event)
            }}>
            <i className={`fa fa-fw ${icon}`}></i>
        </a>
    )
}

Nav.propTypes = {
    flow: PropTypes.object.isRequired,
    active: PropTypes.string.isRequired,
    tabs: PropTypes.array.isRequired,
    onSelectTab: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    onDuplicate: PropTypes.func.isRequired,
    onReplay: PropTypes.func.isRequired,
    onAccept: PropTypes.func.isRequired,
    onRevert: PropTypes.func.isRequired,
}

function Nav({ flow, active, tabs, onSelectTab, onRemove, onDuplicate, onReplay, onAccept, onRevert }) {
    return (
        <nav className="nav-tabs nav-tabs-sm">
            {tabs.map(tab => (
                <a key={tab}
                    href="#"
                    className={classnames({ active: active === tab })}
                    onClick={event => {
                        event.preventDefault()
                        onSelectTab(tab)
                    }}>
                    {_.capitalize(tab)}
                </a>
            ))}
            <NavAction title="[d]elete flow" icon="fa-trash" onClick={() => onRemove(flow)} />
            <NavAction title="[D]uplicate flow" icon="fa-copy" onClick={() => onDuplicate(flow)} />
            <NavAction disabled title="[r]eplay flow" icon="fa-repeat" onClick={() => onReplay(flow)} />
            {flow.intercepted && (
                <NavAction title="[a]ccept intercepted flow" icon="fa-play" onClick={() => onAccept(flow)} />
            )}
            {flow.modified && (
                <NavAction title="revert changes to flow [V]" icon="fa-history" onClick={() => onRevert(flow)} />
            )}
        </nav>
    )
}

export default connect(
    null,
    {
        onRemove: flowsActions.remove,
        onDuplicate: flowsActions.duplicate,
        onReplay: flowsActions.replay,
        onAccept: flowsActions.accept,
        onRevert: flowsActions.revert,
    }
)(Nav)
