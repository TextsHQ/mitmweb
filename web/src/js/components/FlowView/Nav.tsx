import React  from 'react'
import classnames from 'classnames'
import _ from 'lodash'

type NavActionProps = {
    icon: string,
    title: string,
    onClick: (e: any) => void,
}

export function NavAction({ icon, title, onClick }: NavActionProps) {
    return (
        <a title={title}
            href="#"
            className="nav-action"
            onClick={event => {
                event.preventDefault()
                onClick(event)
            }}>
            <i className={`fa fa-fw ${icon}`}></i>
        </a>
    )
}

type NavProps = {
    active: string,
    tabs: string[],
    onSelectTab: (e: string) => void,
}

export default function Nav({ active, tabs, onSelectTab }: NavProps) {
    return (
        <nav className="nav-tabs nav-tabs-sm">
            {tabs.map(tab => (
                <a key={tab}
                    href="#"
                    className={classnames({ active: active === tab })}
                    onClick={event => {
                        event.preventDefault()
                        onSelectTab(tab)
                    }}>
                    {_.capitalize(tab)}
                </a>
            ))}
        </nav>
    )
}
