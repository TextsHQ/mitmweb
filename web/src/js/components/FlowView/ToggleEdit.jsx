import React, { PropTypes, Component } from 'react'
import { connect } from 'react-redux'

import { startEdit, stopEdit } from '../../ducks/ui/flow'

ToggleEdit.propTypes = {
    isEdit: PropTypes.bool.isRequired,
    flow: PropTypes.object.isRequired,
    startEdit: PropTypes.func.isRequired,
    stopEdit: PropTypes.func.isRequired,
}

function ToggleEdit({ isEdit, startEdit, stopEdit, flow }) {
    return (
        <div className="edit-flow-container">
            {isEdit ?
                <a className="edit-flow" onClick={() => stopEdit(flow)}>
                    <i className="fa fa-check"/>
                </a>
                :
                <a className="edit-flow" onClick={() => startEdit(flow)}>
                    <i className="fa fa-pencil"/>
                </a>
            }
        </div>
    )
}

export default connect(
    state => ({
        isEdit: !!state.ui.flow.modifiedFlow,
        flow: state.ui.flow.modifiedFlow || state.flows.byId[state.flows.selected[0]]
    }),
    {
        startEdit,
        stopEdit,
    }
)(ToggleEdit)
