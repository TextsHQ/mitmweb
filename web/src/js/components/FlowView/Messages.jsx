import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'


import { RequestUtils, isValidHttpVersion, parseUrl } from '../../flow/utils.js'
import { formatTimeStamp } from '../../utils.js'
import ContentView from '../ContentView'
import ValidateEditor from '../ValueEditor/ValidateEditor'
import ValueEditor from '../ValueEditor/ValueEditor'

import Headers from './Headers'
import { startEdit, updateEdit, uploadContent } from '../../ducks/ui/flow'
import ToggleEdit from './ToggleEdit'

function RequestLine({ flow, readonly, updateFlow }) {
    return (
        <div className="first-line request-line">
            <div>
                <ValueEditor
                    content={flow.request.method}
                    readonly={readonly}
                    onDone={method => updateFlow({ request: { method } })}
                />
                &nbsp;
                <ValidateEditor
                    content={RequestUtils.pretty_url(flow.request)}
                    readonly={readonly}
                    onDone={url => updateFlow({ request: {path: '', ...parseUrl(url)}})}
                    isValid={url => !!parseUrl(url).host}
                />
                &nbsp;
                <ValidateEditor
                    content={flow.request.http_version}
                    readonly={readonly}
                    onDone={http_version => updateFlow({ request: { http_version } })}
                    isValid={isValidHttpVersion}
                />
            </div>
        </div>
    )
}

function ResponseLine({ flow, readonly, updateFlow }) {
    return (
        <div className="first-line response-line">
            <ValidateEditor
                content={flow.response.http_version}
                readonly={readonly}
                onDone={nextVer => updateFlow({ response: { http_version: nextVer } })}
                isValid={isValidHttpVersion}
            />
            &nbsp;
            <ValidateEditor
                content={flow.response.status_code + ''}
                readonly={readonly}
                onDone={code => updateFlow({ response: { code: parseInt(code) } })}
                isValid={code => /^\d+$/.test(code)}
            />
            &nbsp;
            <ValueEditor
                content={flow.response.reason}
                readonly={readonly}
                onDone={msg => updateFlow({ response: { msg } })}
            />
        </div>
    )
}

const Message = connect(
    state => ({
        flow: state.flows.byId[state.flows.selected[0]],
        isEdit: !!state.ui.flow.modifiedFlow,
    }),
    {
        updateFlow: updateEdit,
        uploadContent: uploadContent
    }
)

export class Request extends Component {
    render() {
        const { flow, isEdit, updateFlow, uploadContent } = this.props

        return (
            <section className="request">
                <ToggleEdit/>
                <RequestLine
                    flow={flow}
                    readonly={!isEdit}
                    updateFlow={updateFlow}/>
                <Headers
                    message={flow.request}
                    readonly={!isEdit}
                    onChange={headers => updateFlow({ request: { headers } })}
                />

                <hr/>
                <ContentView
                    readonly={!isEdit}
                    flow={flow}
                    onContentChange={content => updateFlow({ request: {content}})}
                    uploadContent={content => uploadContent(flow, content, "request")}
                    message={flow.request}/>
            </section>
        )
    }


    edit(k) {
        throw "unimplemented"
        /*
         switch (k) {
         case 'm':
         this.refs.requestLine.refs.method.focus()
         break
         case 'u':
         this.refs.requestLine.refs.url.focus()
         break
         case 'v':
         this.refs.requestLine.refs.httpVersion.focus()
         break
         case 'h':
         this.refs.headers.edit()
         break
         default:
         throw new Error(`Unimplemented: ${k}`)
         }
         */
    }

}

Request = Message(Request)


export class Response extends Component {
    render() {
        const { flow, isEdit, updateFlow, uploadContent } = this.props

        return (
            <section className="response">
                <ToggleEdit/>
                <ResponseLine
                    flow={flow}
                    readonly={!isEdit}
                    updateFlow={updateFlow}/>
                <Headers
                    message={flow.response}
                    readonly={!isEdit}
                    onChange={headers => updateFlow({ response: { headers } })}
                />
                <hr/>
                <ContentView
                    readonly={!isEdit}
                    flow={flow}
                    onContentChange={content => updateFlow({ response: {content}})}
                    uploadContent={content => uploadContent(flow, content, "response")}
                    message={flow.response}
                />
            </section>
        )
    }

    edit(k) {
        throw "unimplemented"
        /*
         switch (k) {
         case 'c':
         this.refs.responseLine.refs.status_code.focus()
         break
         case 'm':
         this.refs.responseLine.refs.msg.focus()
         break
         case 'v':
         this.refs.responseLine.refs.httpVersion.focus()
         break
         case 'h':
         this.refs.headers.edit()
         break
         default:
         throw new Error(`'Unimplemented: ${k}`)
         }
         */
    }
}

Response = Message(Response)


ErrorView.propTypes = {
    flow: PropTypes.object.isRequired,
}

export function ErrorView({ flow }) {
    return (
        <section>
            <div className="alert alert-warning">
                {flow.error.msg}
                <div>
                    <small>{formatTimeStamp(flow.error.timestamp)}</small>
                </div>
            </div>
        </section>
    )
}
