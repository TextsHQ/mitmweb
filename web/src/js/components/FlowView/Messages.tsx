import React from 'react'

import { RequestUtils, isValidHttpVersion, parseUrl } from '../../flow/utils'
import { formatTimeStamp } from '../../utils'
import ContentView from '../ContentView'
import ContentViewOptions from '../ContentView/ContentViewOptions'
import ValidateEditor from '../ValueEditor/ValidateEditor'
import ValueEditor from '../ValueEditor/ValueEditor'
import HideInStatic from '../common/HideInStatic'

import Headers from './Headers'
import { updateEdit as updateFlow } from '../../ducks/ui/flow'
import { uploadContent } from '../../ducks/flows'
import ToggleEdit from './ToggleEdit'
import { useAppDispatch, useAppSelector } from "../../ducks";
import { HTTPFlow, HTTPMessage } from '../../flow'


type RequestLineProps = {
    flow: HTTPFlow,
    readonly: boolean,
}

function RequestLine({ flow, readonly }: RequestLineProps) {
    const dispatch = useAppDispatch()

    return (
        <div className="first-line request-line">
            <div>
                <ValueEditor
                    content={flow.request.method}
                    readonly={readonly}
                    onDone={method => dispatch(updateFlow({ request: { method } }))}
                />
                &nbsp;
                <ValidateEditor
                    content={RequestUtils.pretty_url(flow.request)}
                    readonly={readonly}
                    onDone={url => dispatch(updateFlow({ request: {path: '', ...parseUrl(url)}}))}
                    isValid={url => !!parseUrl(url).host}
                />
                &nbsp;
                <ValidateEditor
                    content={flow.request.http_version}
                    readonly={readonly}
                    onDone={http_version => dispatch(updateFlow({ request: { http_version } }))}
                    isValid={isValidHttpVersion}
                />
            </div>
        </div>
    )
}

type ResponseLineProps = {
    flow: HTTPFlow,
    readonly: boolean,
}

function ResponseLine({ flow, readonly }: ResponseLineProps) {
    const dispatch = useAppDispatch()

    return (
        <div className="first-line response-line">
            <ValidateEditor
                content={flow.response?.http_version}
                readonly={readonly}
                onDone={nextVer => dispatch(updateFlow({ response: { http_version: nextVer } }))}
                isValid={isValidHttpVersion}
            />
            &nbsp;
            <ValidateEditor
                content={flow.response?.status_code + ''}
                readonly={readonly}
                onDone={code => dispatch(updateFlow({ response: { code: parseInt(code) } }))}
                isValid={code => /^\d+$/.test(code)}
            />
            &nbsp;
            <ValueEditor
                content={flow.response?.reason}
                readonly={readonly}
                onDone={msg => dispatch(updateFlow({ response: { msg } }))}
            />
        </div>
    )
}

export function Request() {
    const dispatch = useAppDispatch(),
    flow = useAppSelector(state => state.ui.flow.modifiedFlow || state.flows.byId[state.flows.selected[0]]),
    isEdit = useAppSelector(state => !!state.ui.flow.modifiedFlow)

    let noContent =  !isEdit && (flow.request.contentLength == 0 || flow.request.contentLength == null)
    return (
        <section className="request">
            <article>
                <ToggleEdit/>
                <RequestLine
                    flow={flow}
                    readonly={!isEdit} />
                <Headers
                    message={flow.request}
                    readonly={!isEdit}
                    onChange={headers => dispatch(updateFlow({ request: { headers } }))}
                />

                <hr/>
                <ContentView
                    readonly={!isEdit}
                    flow={flow}
                    onContentChange={content => dispatch(updateFlow({ request: {content}}))}
                    message={flow.request}/>

                <hr/>
                <Headers
                    message={flow.request}
                    readonly={!isEdit}
                    onChange={trailers => dispatch(updateFlow({ request: { trailers } }))}
                    type='trailers'
                />
            </article>
            <HideInStatic>
            {!noContent &&
                <footer>
                    <ContentViewOptions
                        flow={flow}
                        message={flow.request}
                        uploadContent={content => dispatch(uploadContent(flow, content, "request"))}/>
                </footer>
            }
            </HideInStatic>
        </section>
    )
}

export function Response() {
    const dispatch = useAppDispatch(),
    flow = useAppSelector(state => state.ui.flow.modifiedFlow || state.flows.byId[state.flows.selected[0]]),
    isEdit = useAppSelector(state => !!state.ui.flow.modifiedFlow)

    let noContent =  !isEdit && (flow.response.contentLength == 0 || flow.response.contentLength == null)

    return (
        <section className="response">
            <article>
                <ToggleEdit/>
                <ResponseLine
                    flow={flow}
                    readonly={!isEdit} />
                <Headers
                    message={flow.response}
                    readonly={!isEdit}
                    onChange={headers => dispatch(updateFlow({ response: { headers } }))}
                />
                <hr/>
                <ContentView
                    readonly={!isEdit}
                    flow={flow}
                    onContentChange={content => dispatch(updateFlow({ response: {content}}))}
                    message={flow.response}
                />
                <hr/>
                <Headers
                    message={flow.response}
                    readonly={!isEdit}
                    onChange={trailers => dispatch(updateFlow({ response: { trailers } }))}
                    type='trailers'
                />
            </article>
            <HideInStatic>
            {!noContent &&
                <footer >
                    <ContentViewOptions
                        flow={flow}
                        message={flow.response}
                        uploadContent={content => dispatch(uploadContent(flow, content, "response"))} />
                </footer>
            }
            </HideInStatic>
        </section>
    )
}

type ErrorViewProps = {
    flow: HTTPFlow
}

export function ErrorView({ flow }: ErrorViewProps) {
    return (
        <section className="error">
            <div className="alert alert-warning">
                {flow.error?.msg}
                <div>
                    <small>{formatTimeStamp(flow.error?.timestamp)}</small>
                </div>
            </div>
        </section>
    )
}
