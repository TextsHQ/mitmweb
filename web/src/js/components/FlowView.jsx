import React, { Component } from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import Nav from './FlowView/Nav'
import { Request, Response, ErrorView as Error } from './FlowView/Messages'
import Details from './FlowView/Details'
import Prompt from './Prompt'

import { setPrompt, selectTab } from '../ducks/ui'

export default class FlowView extends Component {

    static allTabs = { Request, Response, Error, Details }

    constructor(props, context) {
        super(props, context)
        this.onPromptFinish = this.onPromptFinish.bind(this)
    }

    onPromptFinish(edit) {
        this.props.setPrompt(false)
        if (edit && this.tabComponent) {
            this.tabComponent.edit(edit)
        }
    }

    getPromptOptions() {
        switch (this.props.tab) {

            case 'request':
                return [
                    'method',
                    'url',
                    { text: 'http version', key: 'v' },
                    'header'
                ]
                break

            case 'response':
                return [
                    { text: 'http version', key: 'v' },
                    'code',
                    'message',
                    'header'
                ]
                break

            case 'details':
                return

            default:
                throw 'Unknown tab for edit: ' + this.props.tab
        }
    }

    render() {
        let { flow, tab: active, updateFlow } = this.props
        const tabs = ['request', 'response', 'error'].filter(k => flow[k]).concat(['details'])

        if (tabs.indexOf(active) < 0) {
            if (active === 'response' && flow.error) {
                active = 'error'
            } else if (active === 'error' && flow.response) {
                active = 'response'
            } else {
                active = tabs[0]
            }
        }

        const Tab = FlowView.allTabs[_.capitalize(active)]

        return (
            <div className="flow-detail" onScroll={this.adjustHead}>
                <Nav
                    flow={flow}
                    tabs={tabs}
                    active={active}
                    onSelectTab={this.props.selectTab}
                />
                <Tab ref={ tab => this.tabComponent = tab } flow={flow} updateFlow={updateFlow} />
                {this.state.prompt && (
                    <Prompt options={this.getPromptOptions()} done={this.onPromptFinish} />
                )}
            </div>
        )
    }
}

export default connect(
    state => ({
        needEdit: state.ui.needEdit,
    }),
    {
        setPrompt,
        selectTab,
    }
)(FlowView)
