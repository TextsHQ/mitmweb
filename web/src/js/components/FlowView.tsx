import React from 'react'
import _ from 'lodash'

import Nav from './FlowView/Nav'
import { ErrorView as Error, Request, Response } from './FlowView/Messages'
import Details from './FlowView/Details'

import { selectTab } from '../ducks/ui/flow'
import {useAppDispatch, useAppSelector} from "../ducks";

export const allTabs = { Request, Response, Error, Details }

export default function FlowView() {
    const dispatch = useAppDispatch(),
    flow = useAppSelector(state => state.flows.byId[state.flows.selected[0]])

    let tabName = useAppSelector(state => state.ui.flow.tab)

    // only display available tab names
    const tabs = ['request', 'response', 'error'].filter(k => flow[k])
    tabs.push("details")

    if (tabs.indexOf(tabName) < 0) {
        if (tabName === 'response' && flow.error) {
            tabName = 'error'
        } else if (tabName === 'error' && flow.response) {
            tabName = 'response'
        } else {
            tabName = tabs[0]
        }
    }

    const Tab = allTabs[_.capitalize(tabName)]

    return (
        <div className="flow-detail">
            <Nav
                tabs={tabs}
                active={tabName}
                onSelectTab={(tab: string) => dispatch(selectTab(tab))}
            />
            <Tab flow={flow}/>
        </div>
    )
}
