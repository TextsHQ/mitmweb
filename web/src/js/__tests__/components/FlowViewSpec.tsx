import * as React from "react"
import {render, screen, waitFor} from "../test-utils";
import FlowView from "../../components/FlowView";
import fetchMock, {enableFetchMocks} from "jest-fetch-mock";
import {fireEvent} from "@testing-library/react";

enableFetchMocks();

test("FlowView", async () => {
    fetchMock.mockReject(new Error("backend missing"));

    const {asFragment} = render(<FlowView/>);
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Response"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("WebSocket"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Connection"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Timing"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Error"));
    expect(asFragment()).toMatchSnapshot();
});
