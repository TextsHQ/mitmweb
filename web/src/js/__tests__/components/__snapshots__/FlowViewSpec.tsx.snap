// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowView 1`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class="active"
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="request"
    >
      <div
        class="first-line request-line"
      >
        <div>
          <span
            class="inline-input has-success"
            tabindex="0"
          >
            GET
          </span>
           
          <span
            class="inline-input has-success"
            tabindex="0"
          >
            http://address:22/second
          </span>
           
          <span
            class="inline-input has-success"
            tabindex="0"
          >
            HTTP/1.1
          </span>
        </div>
      </div>
      <div
        class="kv-editor headers"
      >
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            header
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            qvalue
          </span>
        </div>
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            content-length
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            7
          </span>
        </div>
        <div
          aria-label="Add"
          class="kv-add-row fa fa-plus-square-o"
          role="button"
        />
      </div>
      <hr />
      <div
        class="contentview"
      >
        <div
          class="controls"
        >
          <h5>
            Loading...
          </h5>
          <button
            class="btn-xs btn btn-default"
          >
            <i
              class="fa fa-edit"
            />
             Edit
          </button>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
            title="Upload a file to replace the content."
          >
            <i
              class="fa fa-fw fa-upload"
            />
            Replace
            <input
              class="hidden"
              type="file"
            />
          </a>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
          >
            <span>
              <i
                class="fa fa-fw fa-files-o"
              />
               
              <b>
                View:
              </b>
               auto 
              <span
                class="caret"
              />
            </span>
          </a>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 2`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class="active"
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="response"
    >
      <div
        class="first-line response-line"
      >
        <span
          class="inline-input has-success"
          tabindex="0"
        >
          HTTP/1.1
        </span>
         
        <span
          class="inline-input has-success"
          tabindex="0"
        >
          200
        </span>
         
        <span
          class="inline-input"
          tabindex="0"
        >
          OK
        </span>
      </div>
      <div
        class="kv-editor headers"
      >
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            header-response
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            svalue
          </span>
        </div>
        <div
          class="kv-row"
        >
          <span
            class="inline-input kv-key"
            tabindex="0"
          >
            content-length
          </span>
          : 
          <span
            class="inline-input kv-value"
            placeholder="empty"
            tabindex="0"
          >
            7
          </span>
        </div>
        <div
          aria-label="Add"
          class="kv-add-row fa fa-plus-square-o"
          role="button"
        />
      </div>
      <hr />
      <div
        class="contentview"
      >
        <div
          class="controls"
        >
          <h5>
            Loading...
          </h5>
          <button
            class="btn-xs btn btn-default"
          >
            <i
              class="fa fa-edit"
            />
             Edit
          </button>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
            title="Upload a file to replace the content."
          >
            <i
              class="fa fa-fw fa-upload"
            />
            Replace
            <input
              class="hidden"
              type="file"
            />
          </a>
           
          <a
            class="btn btn-default btn-xs"
            href="#"
          >
            <span>
              <i
                class="fa fa-fw fa-files-o"
              />
               
              <b>
                View:
              </b>
               auto 
              <span
                class="caret"
              />
            </span>
          </a>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 3`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class="active"
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="websocket"
    >
      <h4>
        WebSocket
      </h4>
      <div
        class="contentview"
      >
        <div
          class="controls"
        >
          <h5>
            3 Messages
          </h5>
          <a
            class="btn btn-default btn-xs"
            href="#"
          >
            <span>
              <i
                class="fa fa-fw fa-files-o"
              />
               
              <b>
                View:
              </b>
               auto 
              <span
                class="caret"
              />
            </span>
          </a>
        </div>
      </div>
      <div>
        <i
          class="fa fa-fw fa-window-close text-muted"
        />
          Closed by server with code 1000 (Close Reason).
        <small
          class="pull-right"
        >
          1999-12-31 23:00:05.000
        </small>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 4`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class="active"
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="detail"
    >
      <h4>
        Client Connection
      </h4>
      <table
        class="connection-table"
      >
        <tbody>
          <tr>
            <td>
              Address:
            </td>
            <td>
              127.0.0.1:22
            </td>
          </tr>
          <tr>
            <td>
              <abbr
                title="TLS Server Name Indication"
              >
                SNI
              </abbr>
              :
            </td>
            <td>
              address
            </td>
          </tr>
          <tr>
            <td>
              <abbr
                title="ALPN protocol negotiated"
              >
                ALPN
              </abbr>
              :
            </td>
            <td>
              http/1.1
            </td>
          </tr>
          <tr>
            <td>
              TLS Version:
            </td>
            <td>
              TLSv1.2
            </td>
          </tr>
          <tr>
            <td>
              TLS Cipher:
            </td>
            <td>
              cipher
            </td>
          </tr>
        </tbody>
      </table>
      <h4>
        Server Connection
      </h4>
      <table
        class="connection-table"
      >
        <tbody>
          <tr>
            <td>
              Address:
            </td>
            <td>
              address:22
            </td>
          </tr>
          <tr>
            <td>
              Resolved address:
            </td>
            <td>
              192.168.0.1:22
            </td>
          </tr>
          <tr>
            <td>
              Source address:
            </td>
            <td>
              address:22
            </td>
          </tr>
          <tr>
            <td>
              <abbr
                title="TLS Server Name Indication"
              >
                SNI
              </abbr>
              :
            </td>
            <td>
              address
            </td>
          </tr>
          <tr>
            <td>
              TLS Version:
            </td>
            <td>
              TLSv1.2
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 5`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class=""
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class="active"
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="timing"
    >
      <h4>
        Timing
      </h4>
      <table
        class="timing-table"
      >
        <tbody>
          <tr>
            <td>
              Client conn. established:
            </td>
            <td>
              1999-12-31 23:00:00.000
              <span
                class="text-muted"
              >
                (0ms)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              First request byte:
            </td>
            <td>
              1999-12-31 23:00:00.000
            </td>
          </tr>
          <tr>
            <td>
              Client conn. TLS handshake:
            </td>
            <td>
              1999-12-31 23:00:01.000
              <span
                class="text-muted"
              >
                (1s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Request complete:
            </td>
            <td>
              1999-12-31 23:00:01.000
              <span
                class="text-muted"
              >
                (1s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Server conn. initiated:
            </td>
            <td>
              1999-12-31 23:00:02.000
              <span
                class="text-muted"
              >
                (2s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              First response byte:
            </td>
            <td>
              1999-12-31 23:00:02.000
              <span
                class="text-muted"
              >
                (2s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Server conn. TCP handshake:
            </td>
            <td>
              1999-12-31 23:00:03.000
              <span
                class="text-muted"
              >
                (3s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Response complete:
            </td>
            <td>
              1999-12-31 23:00:03.000
              <span
                class="text-muted"
              >
                (3s)
              </span>
            </td>
          </tr>
          <tr>
            <td>
              Server conn. TLS handshake:
            </td>
            <td>
              1999-12-31 23:00:04.000
              <span
                class="text-muted"
              >
                (4s)
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</DocumentFragment>
`;

exports[`FlowView 6`] = `
<DocumentFragment>
  <div
    class="flow-detail"
  >
    <nav
      class="nav-tabs nav-tabs-sm"
    >
      <a
        class=""
        href="#"
      >
        Request
      </a>
      <a
        class=""
        href="#"
      >
        Response
      </a>
      <a
        class=""
        href="#"
      >
        WebSocket
      </a>
      <a
        class="active"
        href="#"
      >
        Error
      </a>
      <a
        class=""
        href="#"
      >
        Connection
      </a>
      <a
        class=""
        href="#"
      >
        Timing
      </a>
    </nav>
    <section
      class="error"
    >
      <div
        class="alert alert-warning"
      >
        error
        <div>
          <small>
            1999-12-31 23:00:07.000
          </small>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
